### PR TITLE
Equal for SymbolicOp subclasses

### PR DESCRIPTION
**Pull request by lillian542**
**Context:**
The `qml.equal` function cannot compare subclasses of `SymbolicOp` (`Adjoint`, `SProd`, `Exp`, `Pow`).

**Description of the Change:**
Now it can.

**Possible Drawbacks:**
If `arithmetic_depth` is greater than 1, and operators are nested the same way, it can compares them correctly. If they are nested in a different order, it returns False, even if the order of the nesting is not significant. For example:

```
nested_op = qml.pow(qml.pow(base, 2), 3)
nested_op2 = qml.pow(qml.pow(base, 3), 2)

qml.equal(nested_op, nested_op) 
>>> True

qml.equal(nested_op, nested_op2)
>>> False
```
Not sure if it would be better to raise a `NotImplementedError` for `arithmetic_depth` greater than 1, or to leave this (since it can correctly identify that identical operators are equal) and note in the docstring that the function assumes a mismatch in the nesting order means the operators aren't equivalent.

### CODE REVIEW

File: pennylane/ops/functions/equal.py [hunk 1]
```diff
@@ -223,6 +223,41 @@ def _equal_controlled(op1: Controlled, op2: Controlled, **kwargs):
         ) from e
 
 
+@_equal.register
+# pylint: disable=unused-argument
+def _equal_pow(op1: Pow, op2: Pow, **kwargs):
```
Comment (antalszava) [inline pennylane/ops/functions/equal.py:228]: As we chatted, maybe adding a note on the fact that the attributes of the operators are being checked could be useful just to help users.

File: tests/ops/functions/test_equal.py [hunk 2]
```diff
@@ -1238,46 +1290,40 @@ def test_control_wires_comparison(self, wire1, wire2, res):
...
-        import torch
-        import jax
-
-        base1 = qml.RX(torch.tensor(1.2), wires=0)
-        base2 = qml.RX(jax.numpy.array(1.2), wires=0)
-
-        op1 = Controlled(base1, control_wires=1)
-        op2 = Controlled(base2, control_wires=1)
+    @pytest.mark.parametrize("base", PARAMETRIZED_OPERATIONS)
+    def test_adjoint_comparison(self, base):
```
Comment (antalszava) [inline tests/ops/functions/test_equal.py:1302]: Test docstrings seem to be missing here.

File: tests/ops/functions/test_equal.py [hunk 3]
```diff
@@ -1199,35 +1203,83 @@ def test_unsupported_object_type_not_implemented(self):
...
+        (qml.PauliX(0), qml.PauliX(0), True),
+        (qml.PauliX(0) @ qml.PauliY(1), qml.PauliX(0) @ qml.PauliY(1), True),
+        (qml.CRX(1.23, [0, 1]), qml.CRX(1.23, [0, 1]), True),
+        (qml.CRX(1.23, [1, 0]), qml.CRX(1.23, [0, 1]), False),
+        (qml.PauliY(1), qml.PauliY(0), False),
+        (qml.PauliX(1), qml.PauliY(1), False),
+        (qml.PauliX(0) @ qml.PauliY(1), qml.PauliZ(1) @ qml.PauliY(0), False),
+    ]
+
+    PARAMS = [(1.23, 1.23, True), (5, 5, True), (2, -2, False), (1.2, 1, False)]
```
Comment (antalszava) [inline tests/ops/functions/test_equal.py:1219]: Nice to parametrize these! :clap: :slightly_smiling_face:

File: tests/ops/functions/test_equal.py [hunk 4]
```diff
@@ -1199,35 +1203,83 @@ def test_unsupported_object_type_not_implemented(self):
...
+
+    def test_mismatched_arithmetic_depth(self):
```
Comment (antalszava) [inline tests/ops/functions/test_equal.py:1221]: Just double-checking: these tests were moved in the file untouched, correct?

**Review by antalszava — APPROVED** [review]
Looks good! :slightly_smiling_face: It's cool that `qml.equal` is becoming more and more powerful. :tada:

File: tests/ops/functions/test_equal.py (hunk 4 above)
Comment (lillian542) [reply tests/ops/functions/test_equal.py:1221]: Yes, just copy-pasted

File: tests/ops/functions/test_equal.py (hunk 2 above)
Comment (lillian542) [reply tests/ops/functions/test_equal.py:1302]: Thanks, added now

**Pull request merged**